### PR TITLE
2 different Peierls Creep rheologies and Grain Boundary Sliding (GBS)

### PR DESCRIPTION
**Pull request by Dr3ven**
I implemented 2 different Peierls creep laws. The NonLinearPeierlsCreep has a power law part in its equation, which makes it differ from the "normal" Peierls creep law that is implemented.
Additionally, corresponding creep laws are implemented for both new rheologies.

### CODE REVIEW

File: test/test_GrainBoundarySliding.jl [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    εII = zero(τII)
+    compute_εII!(εII, p, τII, args)
+    eta_array1 = @. 0.5 * τII / εII
+
+    # test overriding the default values
+    a =  SetGrainBoundarySliding("Dry Olivine >= 1523K | Hirth and Kohlstedt (2003)", V=1e-6m^3/mol)
+    @test Value(a.V) == 1.0e-6m^3/mol
+
+
+    # --- debugging
```
Comment (albert-de-montserrat) [inline test/test_GrainBoundarySliding.jl:77]: Remove commented lines

File: test/test_NonLinearPeierlsCreep.jl [hunk 2]
```diff
@@ -0,0 +1,102 @@
+using Test
+using GeoParams
+
+@testset "NonLinearPeierlsCreepLaws" begin
```
Comment (albert-de-montserrat) [inline test/test_NonLinearPeierlsCreep.jl:4]: Remove all the commented lines of code

File: test/test_PeierlsCreep.jl [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    args_array = (; T=T_array)
+
+    compute_εII!(ε_array, p, τII_array, args_array)
+    @test ε_array[1] ≈ ε
+
+    # compute when args are scalars
+    compute_εII!(ε_array, p, τII_array, args)
+    @test ε_array[1] ≈ ε
+
+    # ===
```
Comment (albert-de-montserrat) [inline test/test_PeierlsCreep.jl:52]: Remove

File: test/test_PeierlsCreep.jl [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+        # Check that values are the same after non-dimensionalisation & dimensionalisation
+        for field in fieldnames(typeof(p_dim))
+            val_original = getfield(p,    field)
+            val_final    = getfield(p_dim,field)
+            if isa(val_original, GeoUnit)
+                @test Value(val_original) == Value(val_final)        
+            end
+        end
+############################## Ab hier nochmal gucken ######################################
```
Comment (albert-de-montserrat) [inline test/test_PeierlsCreep.jl:100]: German

**Review by albert-de-montserrat — CHANGES_REQUESTED** [review]

File: src/CreepLaw/DiffusionCreep.jl [hunk 5]
```diff
@@ -232,7 +232,9 @@ returns the derivative of strainrate versus stress
     @unpack_val n, r, p, A, E, V, R = a
     FT, FE = a.FT, a.FE
 
-    return fastpow(FT * TauII, -1 + n) *
+    # computed symbolically
+    return n *
```
[the remark this answers is not in the document]
Comment (Dr3ven) [reply src/CreepLaw/DiffusionCreep.jl:236]: There is one diffusion creep law that we have implemented (Faul & Jackson, 2006) for dry olivine that has n=1.4.

File: src/CreepLaw/NonLinearPeierlsCreep.jl [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+) where {N,_T}
+    @inbounds for i in eachindex(EpsII)
+        EpsII[i] = compute_εII(a, TauII[i]; T=T[i])
+    end
+
+    return nothing
+end
+
+dεII_dτII(a::NonLinearPeierlsCreep, TauII; args...) = ForwardDiff.derivative(x -> compute_εII(a, x; args...), TauII)
+
```
[the remark this answers is not in the document]
Comment (Dr3ven) [reply src/CreepLaw/NonLinearPeierlsCreep.jl:200]: Boris suggested to iterate for the non-linear solution of the power law stress equation with the local_iterations_εII() function in CompositeRheologies/NonlinearIterations.jl. That should be an option for the commented out stress calculation function in this case, right?

File: src/CreepLaw/PeierlsCreep.jl [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    compute_τII(a::PeierlsCreep, EpsII; P, T, f, args...)
+
+Computes the stress for a peierls creep law given a certain strain rate.
+
+"""
+@inline function compute_τII(
+    a::PeierlsCreep, EpsII::_T; T=one(precision(a)), args...
+) where {_T}
+    local n, q, o, TauP, A, E, R
+    if EpsII isa Quantity
```
[the remark this answers is not in the document]
Comment (Dr3ven) [reply src/CreepLaw/PeierlsCreep.jl:235]: I can do that by looking that it allocates 0.0 space, correct?

File: src/Plotting/test_script.jl [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+#x4 = SetDiffusionCreep("Dry Olivine | Hirth & Kohlstedt (2003)")
+x4 = SetDiffusionCreep("TEST DIFF")
+#x4 = SetDiffusionCreep("THORSTEN DIFF")
+x4 = remove_tensor_correction(x4)
+c2 = CompositeRheology(x1, x2, x3, x4)
+
+# angepasste values damit der plot UNGEFÄHR passt. rausbekommen wie man von ursprünglichen values auf angepasste values kommt und warum.
+
+args = (P=1.0e9,)
+#args = (P=1.0e9, d=3.0)
```
[the remark this answers is not in the document]
Comment (Dr3ven) [reply src/Plotting/test_script.jl:44]: I will remove the whole script. It was just to test some stuff and does not belong in the package.

File: src/CreepLaw/NonLinearPeierlsCreep.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dr3ven) [reply src/CreepLaw/NonLinearPeierlsCreep.jl:201]: Update: Stress can now be calculated using the `Peierls_stress_iteration()` function with Real numbers.

File: src/CreepLaw/PeierlsCreep.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dr3ven) [reply src/CreepLaw/PeierlsCreep.jl:235]: Update: Checked it. It is type stable.

**Comment by Dr3ven** [comment]
I fixed the missing compilation errors. Just the ones left for the 1.10.0 versions and the nightly. For those I don't know what and how to fix it though.

**Comment by albert-de-montserrat** [comment]
Great thanks! Don't worry about 1.10 and nightly for now 👍

File: src/CreepLaw/Data/NonLinearPeierlsCreep.jl [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                year={2010},
+                publisher={Wiley Online Library}
+                }
+        "),
+    ),
+)
+
+# Wet Olivine rheology of Mei 2010 TEST
+(
+    "TEST PEIERLS",
```
Comment (albert-de-montserrat) [inline src/CreepLaw/Data/NonLinearPeierlsCreep.jl:45]: Name not in upper case

File: src/CreepLaw/Data/NonLinearPeierlsCreep.jl [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        ),
+        MaterialParamsInfo(;
+            Comment="Law to check Thorsten Beckers book rheology and its conversion of the A factor",
+            BibTex_Reference=""),
+    ),
+)
+
+# Wet Olivine rheology of Mei 2010 TEST
+(
+    "THORSTEN PEIERLS",
```
Comment (albert-de-montserrat) [inline src/CreepLaw/Data/NonLinearPeierlsCreep.jl:65]: Again, not in upper case

File: src/CreepLaw/DiffusionCreep.jl [hunk 11]
```diff
@@ -232,7 +232,9 @@ returns the derivative of strainrate versus stress
     @unpack_val n, r, p, A, E, V, R = a
     FT, FE = a.FT, a.FE
 
-    return fastpow(FT * TauII, -1 + n) *
+    # computed symbolically
```
Comment (albert-de-montserrat) [inline src/CreepLaw/DiffusionCreep.jl:235]: Could you test if computing this derivative with `ForwardDiff.jl` is actually faster than the symbolic derivative?

File: src/CreepLaw/PeierlsCreep.jl [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+           (FE * R * T * TauII)
+end
+
+@inline function dεII_dτII(
+    a::PeierlsCreep, TauII::Quantity; T=1K, args...
+)
+    @unpack_units n, q, o, TauP, A, E, R = a
+    FT, FE = a.FT, a.FE
+
+    return o *
```
Comment (albert-de-montserrat) [inline src/CreepLaw/PeierlsCreep.jl:214]: Check here as well if ForwardDiff is faster than the symbolic expression.

**Review by albert-de-montserrat — CHANGES_REQUESTED** [review]

File: src/CreepLaw/PeierlsCreep.jl (hunk 12 above)
Comment (Dr3ven) [reply src/CreepLaw/PeierlsCreep.jl:214]: It seems that the symbolic expression is faster:
```julia
# symbolic
@btime dεII_dτII($x1, $TauII, $args)
  62.567 ns (0 allocations: 0 bytes)
3.2013187520023984e-55
 
 # ForwardDiff
 @btime dεII_dτII($x1, $TauII, $args)
  70.255 ns (0 allocations: 0 bytes)
3.2078554209110495e-55
 ```

**Comment by albert-de-montserrat** [comment]
Ready to be merged then?

**Review by albert-de-montserrat — APPROVED** [review]

**Comment by albert-de-montserrat** [comment]
Thanks @Dr3ven !

**Comment by Dr3ven** [comment]
Sure!

**Pull request merged**